### ISSUE TITLE
Question about data scaling

### ISSUE THREAD

**Issue by reeceomahoney** [intuitive-robots/beso#3]
Hi, I've been playing around with this repo and noticed that, at least for my datasets, using the regular scaler completely breaks the model, even if sigma data is adjusted to 1. Additionally, the sigma data produced by the min max scaler is closer to 0.33 but you have chosen 0.5. I was wondering if you could explain your thought process behind the choice of the scaler and sigma data value?

**Comment by mbreuss** [comment]
Hi,

Thanks for your interest in this repo. Sorry for the late reply, I just noticed your post. 
My choice of $\sigma_{\text{data}}$ das was based on the edm framework from: https://arxiv.org/pdf/2206.00364 
I never assumed, that it was related to the std of the underlying training data distribution, which makes a lot of sense after thinking about it now. Upon reading your comment I  read the edn paper again and its mentioned in one of the first parargraphs. 
Back then I was trying to find a good set of  hyperparameters for different settings, that work without changes and just set a few parameters constant including $\sigma_{\text{data}}=0.5$, because it was hard to tune all them in parallel.
I guess I have to experiment  now more with  $\sigma_{\text{data}}$ depending on the actual action distribution.
Thank you for pointing that out!

By regular scaler do you mean zero mean unit variance or the MinMax Scaler in range of [-1,1]?
The Zero-Mean scaler does not work well for EDM at all, I can only recommend MinMax. Or did I missunderstood your question?

**Comment by mbreuss** [comment]
As a general tipp for hyperparameter choice: 

For my current work, I am still using the same Beso Framework with $[-1, 1]$ MinMax Action Scaling and DDIM Sampling, but I found, that with LogLogistic Training Sampling the noise range of $[0.001,80]$ to work well in all of my new experiments in simulation and real world with action-chunking diffusion policies. I tested it with the block push and relay kitchen here, but somehow it does not work as well in these two environments.
But I recommend these values as a starting point, that I had a lot of success with: https://openreview.net/pdf?id=Pt6fLfXMRW

Hope that helps!

**Comment by reeceomahoney** [comment]
Thanks for the reply :) Good to know the Gaussian scaler being broken isn't just me doing something wrong. Regarding the log logistic parameters, I'm assuming you use the same max to initialise your sampling. 80 seems much higher than what I've been using ( 1 ), and use used previously but you're saying your results show that it works well?

**Comment by mbreuss** [comment]
Hi, 
yes it works well with the noise range of [0.001,80] which is the standard for most edm diffusion image generation models such as the new edm2 and more. I tried to make it work with this range to have the chance to work on unified models in the future. However, all new models I am using are either predicting 10 actions at once or 20 actions while BESO does single action prediction. One guess of mine is that higher noise levels work better with higher denoising dimensions. I tested the stuff in 1D toy takes in my consistency Repo and there 80 is also not ideal for 1D stuff and noise ranges of [0.01,1] work better there too. However this is just my experience without proper analysis of this issue.

**Comment by reeceomahoney** [comment]
Ok nice I tried this too and it really improved my results. One last question, did your classifier guided sampler code work? I need to use it for my own work, wondering if you had any success with it?

**Comment by mbreuss** [comment]
Happy to hear that! Looking forward to your project, let me know when its published. 
Regarding CFG, yes it works in this repo. I have trained cfg policies for both environments, where you can control the amount of goal guidance for the model. I created wrapper for sampling. However compared to default models, the models usually performed a bit worse than default goal-conditioned ones. 
I summarized my findings in the paper experiment section.

**Comment by reeceomahoney** [comment]
Oh sorry, I meant classifier guidance, not classifier-free. You have code for both in your project. Did the classifier guidance code work?

**Comment by mbreuss** [comment]
So the code should work, since its adopted form image-based codebase, where classifier guidance was used. I never worked a lot with Classifier-Guidance myself, since you need an action gradient, that is related to your goal, which did not make a lot of sense in my project to learn $f(g,a)$ with some additional model. However, given the success of this method in image generation and some successful usages in RL it should work here too. 
I experimented with other ideas to combine ebms with diffusion, which was not so successful. 
There are some recent papers, which maybe could be relevant for your setting: 

https://arxiv.org/pdf/2404.14743
https://arxiv.org/pdf/2404.06356v1
